### PR TITLE
Make `Null` a subclass of `AnyVal` under `-Yexplicit-nulls`.

### PR DESCRIPTION
**Pull request by sjrd**
Quick! Open this before the new policy requires me to discuss it first. :see_no_evil: (Seriously, if it's not shut down by Core, I'll open a Contributors thread before this moves any further. I just want to see what the CI has to say about it.)

---

As always, no AI here.

---

That removes it from the top of the hierarchy, removing a special-case in the type system. We restore the property that Scala value types are partitioned into `AnyVal` and `AnyRef`.

It makes sense because it has all the properties of other `AnyVal`s: "two" instances of `Null` "re-created" from the keyword `null` are indistinguishible from each other. So it has structural equality.

This does not change the erasure of any type, because `AnyVal` erases to `Object` anyway.

If you have trouble wrapping your head around this, observe that `null: Null` and `(): Unit` have a lot in common. They are both types with a singleton instance, accessible with syntax. `null` is only `== null`, and `()` is only `== ()`. Moreover, they do not have the `eq` operator.

### CODE REVIEW

**Comment by mbovel** [comment]
Is that it? 😄 

image

Source: https://docs.scala-lang.org/scala3/reference/experimental/explicit-nulls.html.

**Comment by sjrd** [comment]
That's the image, yes, but I want the source of the image to redraw the graph in the correct need way. Not patch it up on top. 😉

**Comment by SolalPirelli** [comment]
@sjrd Just remind him he has a paper draft to write ;)

File: library/src/scala/collection/mutable/AnyRefMap.scala
Comment (hamzaremmal) [inline library/src/scala/collection/mutable/AnyRefMap.scala:1]: @sjrd was this change meant to be included in this PR? (Same for the change in `LongMap`)

**Comment by hamzaremmal** [comment]
Following up on https://github.com/scala/scala3/pull/25399#issuecomment-3974060473 here.

@sjrd How exactly does the fact that we compile `AnyVal` affect this PR? So far, it seems we are heading towards a green PR.

**Comment by sjrd** [comment]
See this paragraph on the commit message: 

> Unfortunately, this requires complicating the loading mechanism
for `scala.Null`. It now extends a class that can be compiled
from source, and that is incompatible with it being a "core"
class that is forced in `Definitions.init()`. We make it a
synthesized class instead, like the magic `FunctionN` classes.
I'm not happy about this. The silver lining is that, once we
actually move all-in to explicit-nulls, we will be able to define
`Null` itself in source, which will remove all of that magic.

File: library/src/scala/collection/mutable/AnyRefMap.scala
Comment (hamzaremmal) [reply library/src/scala/collection/mutable/AnyRefMap.scala:1]: Maybe it had to do with this [PR](https://github.com/scala/scala3/pull/25367) 😄

**Comment by SethTisue** [comment]
if #23948 goes through that will provide some additional assurance

**Comment by SethTisue** [comment]
as I said at core today, it does seem surprising for `Null` to no longer live under `AnyRef`. but I don't think it's any _more_ surprising to move it under `AnyVal` instead of `Matchable` (and the latter is already the status quo under explicit nulls)

**Comment by sjrd** [comment]
Contributors thread: https://contributors.scala-lang.org/t/make-null-a-subclass-of-anyval-under-yexplicit-nulls/7406

**Comment by alexandru** [comment]
I don’t think that should happen, because then we’d have this being a legal assignment and it feels wrong:

```scala
val x: Int | Null = null
val y: AnyVal = x
```

The premise is that `Null` should be treated as `Unit`, however, they are not the same. For one, `Unit` does not show up in types like this very often, as it doesn’t make much sense in code, even if it’s legal:

```scala
val x: Int | Unit = ()
```

This is also why the compiler has different behavior when it comes to type-inference. When it comes to *untagged union types*, being specific when `Null` is involved matters, hence:

```scala
scala> val x = if (2 == 1+1) 42 else null
val x: Int | Null = 42

scala> val x = if (2 == 1+1) 42
scala> :type x
Unit
```

Considering the proposed changes, here:
```scala
scala> val x: AnyVal = 10
val x: AnyVal = 10

scala> val y = if (2==1+1) x else 42.0
val y: AnyVal = 10

// What happens here if the proposed change comes through?
scala> val y = if (2==1+1) x else null
val y: AnyVal | Null = 10
```

Note that in Kotlin,  the language has `Any` versus `Any?`.

I'm out of my debt here, but one has to consider future evolutions of the language, and I can think of 2 possible issues:

- Unrestricted untagged union types probably slow down the compiler, which is why the Kotlin language, in introducing [Rich Errors](https://github.com/Kotlin/KEEP/blob/main/proposals/KEEP-0441-rich-errors-motivation.md#type-hierarchy) (see [presentation](https://youtu.be/IUrA3mDSWZQ)), abstains from introducing generalized untagged union types, with `Error` being a parallel type hierarchy to `Any` and `Any?`; Kotlin's errors can't be generic at this point either; the reason given is to prevent exponential complexity in the type inference algorithm. That boat has sailed for Scala 3, obviously, and I like that BTW, but the language may still choose to be conservative in what it infers in the future, and inferring `AnyVal` instead of `AnyVal | Null` sounds bad.
- It may not seem significant now, as basically the only `AnyVal` values we actually work with are `Char`, `Int`, `Long` and `Double`, but value classes are coming to the JVM, and again, I'd never want to see `AnyVal` instead of `AnyVal | Null`.

I'm not a compiler expert, hope this criticism is constructive, sorry if this adds noise.

**Comment by sjrd** [comment]
Would you like posting that feedback on the Contributors thread instead? Let's keep the discussion about *whether* to do it on Contributors, and the discussion on *how* to do it on the PR.

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 1]
```diff
@@ -477,13 +478,22 @@ class Definitions {
...
-  }
+  @tu lazy val NullClass: ClassSymbol = requiredClass("scala.Null")
   def NullType: TypeRef = NullClass.typeRef
 
+  private def synthesizeNullClass(): ClassSymbol = {
+    // When explicit-nulls is enabled, Null becomes a direct subtype of AnyVal
+    val completer = new LazyType {
+      def complete(denot: SymDenotation)(using Context): Unit = {
+        val cls = denot.asClass.classSymbol
+        val parents = if ctx.explicitNulls then AnyValType :: Nil else ObjectType :: Nil
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:500]: Here's a controversial suggestion on a controversial PR:
```suggestion
        val parents = AnyValType :: Nil
```

TypeComparer already has logic that makes Null <: T for any T of reference type. In particular, this includes when T = AnyRef (ObjectType).

The logic in TypeComparer is gated by the SafeNulls mode and can therefore be scoped under a language import. The SafeNulls mode is itself gated under -Yexplicit-nulls.

Thus, if we applied this suggestion, we would have:

- Under UnsafeNulls or when -Yexplicit-nulls is disabled: Null <: AnyVal | AnyRef
- Under SafeNulls when -Yexplicit-nulls is enabled: Null <: AnyVal

The point is that the relationship would then be scoped under a language import rather than global decision.

I'm very curious about @sjrd 's thoughts on this.
cc @odersky @noti0na1

Comment (sjrd) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:500]: That would require a SIP independent of the explicit-nulls feature. It's not worth it.

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:500]: OK, makes sense.

**Comment by sjrd** [comment]
This is now ready for review.

**Comment by sjrd** [comment]
HTML Proofer is failing, but otherwise green. Is this spurious?

**Comment by noti0na1** [comment]
Can you add some tests about pattern matching `AnyVal` with and without `null`? Not sure if we should get any warning about exhaustivity.

File: tests/explicit-nulls/neg/basic.scala [hunk 2]
```diff
@@ -32,5 +32,5 @@ class Basic {
   val c2: Int = n // error
   val i3: Int | Null = null
 
-  val av: AnyVal = null // error
+  val av: AnyVal = null // ok
```
Comment (noti0na1) [inline tests/explicit-nulls/neg/basic.scala]: Let's add `val x: AnyVal = (??? : Int | Null)` as well

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 3]
```diff
@@ -2192,7 +2202,6 @@ class Definitions {
       orType,
       RepeatedParamClass,
       ByNameParamClass2x,
-      NullClass,
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:2195]: Why do we remove it from this list?

Comment (sjrd) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:2195]: Because it extends `AnyVal`, which is loaded from TASTy. Classes in this list are loaded before any TASTy gets any chance to be read, so a class that extends `AnyVal` cannot be in that list.

In turn, because it can't be in that list anymore, we need to synthesize it on-demand in the definition of `ScalaPackageClass`.

See also the commit message.

**Review by sjrd — COMMENTED** [review]
Rebased and added the requested test.

**Review by noti0na1 — APPROVED** [review]
